### PR TITLE
feedback: Extend features report

### LINKED ISSUE #3608: OS version and containerization detection in feedback daemon
Please add detection of OS version and detection of Docker to the feedback daemon.
This will help us prioritize platform support in future releases

### PR DESCRIPTION
**Pull request by RunsFor**
This patchset adds more information about running
tarantool instance. Firstly, it collects **platform information**
such as os, arch and cgroup environment. Secondly, it looks
through box user spaces and collects info about **db engines and
indices**.

Example of full feedback report:

```json
{
  "arch": "x64",
  "features": {
    "schema": {
      "local_spaces": 0,
      "functional_indices": 0,
      "hash_indices": 0,
      "rtree_indices": 0,
      "temporary_spaces": 0,
      "tree_indices": 0,
      "jsonpath_indices": 0,
      "vinyl_spaces": 0,
      "bitset_indices": 0,
      "memtx_spaces": 0
    }
  },
  "server_id": "79047619-2c60-4195-af4f-23dee72285e6",
  "cgroup": "",
  "os": "OSX",
  "cluster_id": "57da2972-c33f-4b68-8e98-12cffb2fe16f",
  "tarantool_version": "2.5.0-147-g4a80bc779",
  "feedback_version": 2
}
```

Closes #3608
Part of #4943

### CODE REVIEW

File: src/box/lua/feedback_daemon.lua [hunk 1]
```diff
@@ -26,13 +26,40 @@ local function get_fiber_id(f)
     return fid
 end
 
-local function fill_in_feedback(feedback)
+local function detect_docker_environment()
+    local fh = fio.open('/proc/1/cgroup', {'O_RDONLY'})
```
Comment (olegrok) [inline src/box/lua/feedback_daemon.lua:30]: But you could just check `/.dockerenv` file existence.

Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua:30]: I suppose, this file may be removed any time soon. It was originally used by lxc execution driver, which was removed from docker 4 years ago. See more [here](https://superuser.com/a/1021925). I would rather stick with "reading /proc/1/cgroup file" approach

File: src/box/lua/feedback_daemon.lua [hunk 2]
```diff
@@ -26,13 +26,108 @@ local function get_fiber_id(f)
...
+    fh:close()
+    if not ok then return false end
+
+    if not string.find(ok, 'docker') then return false end
+
+    return true
+end
+
+local BOX_SYSTEM_ID_MIN = 256
+local BOX_SYSTEM_ID_MAX = 511
```
Comment (olegrok) [inline src/box/lua/feedback_daemon.lua:45]: box.schema.SYSTEM_ID_MIN and box.schema.SYSTEM_ID_MAX

Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: Resolved, thanks

**Comment by sharonovd** [comment]
I'd suggest to use counters instead of booleans, like this:
temporary_spaces: 5
vinyl_spaces: 3

etc.

File: src/box/lua/feedback_daemon.lua [hunk 3]
```diff
@@ -26,13 +26,105 @@ local function get_fiber_id(f)
...
+
+    -- fh:read() doesn't read empty files
+    local big_enough_chunk = 4096
+    local ok = fh:read(big_enough_chunk)
+    fh:close()
+    if not ok then return false end
+
+    if not string.find(ok, 'docker') then return false end
+
+    return true
```
Comment (printercu) [inline src/box/lua/feedback_daemon.lua:41]: `return str and str:find('docker') or false`?

File: src/box/lua/feedback_daemon.lua [hunk 4]
```diff
@@ -26,13 +26,105 @@ local function get_fiber_id(f)
...
+end
+
+local function is_system_space(sp)
+    local sp_id = sp.id
+    if box.schema.SYSTEM_ID_MIN <= sp_id and sp_id <= box.schema.SYSTEM_ID_MAX then
```
Comment (printercu) [inline src/box/lua/feedback_daemon.lua:46]: `return box.schema.SYSTEM_ID_MIN <= sp_id and sp_id <= box.schema.SYSTEM_ID_MAX` ?

File: src/box/lua/feedback_daemon.lua [hunk 5]
```diff
@@ -26,13 +26,105 @@ local function get_fiber_id(f)
...
+            end
+            if sp.is_local ~= nil then is_local = true end
+            fill_in_space_indices(feedback, sp)
+        end
+    end
+
+    feedback.features.has_temporary_spaces = is_temporary
+    feedback.features.has_local_spaces = is_local
+    feedback.features.memtx = is_memtx
+    feedback.features.vinyl = is_vinyl
```
Comment (printercu) [inline src/box/lua/feedback_daemon.lua:120]: Should it be `has_memtx` and `has_vinyl` as well?

File: src/box/lua/feedback_daemon.lua [hunk 6]
```diff
@@ -26,13 +26,105 @@ local function get_fiber_id(f)
...
+
+    if idx_count > 1 then
+        feedback.features.has_secondary_index = true
+    end
+end
+
+local function fill_in_features(feedback)
+    feedback.features = feedback.features or {}
+
+    local is_memtx, is_vinyl, is_temporary, is_local
```
Comment (printercu) [inline src/box/lua/feedback_daemon.lua:103]: Let's use `has_` so names match fields in the result.

File: src/box/lua/feedback_daemon.lua (hunk 3 above)
Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: thanks, actually it's:

```lua
return str and str:find('docker') and true or false
```

File: src/box/lua/feedback_daemon.lua (hunk 4 above)
Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: done

File: src/box/lua/feedback_daemon.lua (hunk 5 above)
Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: I moved from flags to counters, so now it has different names. See updated PR description

File: src/box/lua/feedback_daemon.lua (hunk 6 above)
Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: No more flags in the latest version

**Comment by RunsFor** [comment]
> I suggest using space instead of sp. Especially in function arg names.

@printercu Why? When the context is short, I think it is ok:

```lua
local function is_system_space(sp)
    return box.schema.SYSTEM_ID_MIN <= sp.id and
            sp.id <= box.schema.SYSTEM_ID_MAX
end
```

Also, it avoids a tautology like `local function is_system_space(space) ... end`

**Comment by RunsFor** [comment]
> I'd suggest to use counters instead of booleans, like this:
> temporary_spaces: 5
> vinyl_spaces: 3
> 
> etc.

Thanks for your suggestion. I've moved flags to counters in the latest update

**Comment by printercu** [comment]
> Why? When the context is short, I think it is ok

Context for this function is a whole file. I didn't immediately get that `sp` stays for `space`. Using understandable names makes code more self-documented, otherwise more comments are required (like `@param sp - space to check`). For 2 line function perception improvement may be not so obvious, but having a rule "For functions longer than 5 lines use accurate names, otherwise it's allowed to use obfuscated ones" is complicated and looks strange :) So if it's better for long functions why not to use it for all?

And I think there is no need to worry about tautology, because it's common in procedural code: context is shared between very different functions and it's ok to clarify the context in function names. And the function signature is the only place where argument names goes along with function name: all other code uses function name with local names for values to pass in arguments.

File: src/box/lua/feedback_daemon.lua [hunk 7]
```diff
@@ -26,13 +26,157 @@ local function get_fiber_id(f)
...
+end
+
+local function is_system_space(sp)
+    return box.schema.SYSTEM_ID_MIN <= sp.id and
+            sp.id <= box.schema.SYSTEM_ID_MAX
+end
+
+local function is_jsonpath_index(idx)
+    for _, part in pairs(idx.parts) do
+        if part.path ~= nil then
```
Comment (olegrok) [inline src/box/lua/feedback_daemon.lua:60]: Such approach doesn't distinguish jsonpath indexes from multikey indexes. It's a bit different things.

Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: Done. Do not increment a counter if `path` contains `[*]` substring

Comment (olegrok) [reply src/box/lua/feedback_daemon.lua]: I suppose you should add separate metric for multikey indexes. E.g. why are functional indexes better than multikey (you count functional indexes but skip multikey)?

Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: I agree. There is also functional multikey index, which is also worth count

Comment (Totktonada) [reply src/box/lua/feedback_daemon.lua]: Multikey indices should be considered as jsonpath indices, IMHO.

Comment (RunsFor) [reply src/box/lua/feedback_daemon.lua]: @olegrok Added multikey metric for both jsonpath and functional indices.
@Totktonada Include multikey indices in jsonpath counter. Same for functional multikey indices.

**Comment by kyukhin** [comment]
Looks like it is dead.

**Pull request closed without merging**